### DISCUSSION TITLE
Retrieve prompts based on versioning rather than times for caching

### DISCUSSION THREAD

**Discussion by ajram23** [langfuse/langfuse#1904] [Ideas]
### Describe the feature or potential improvement

Prompts don't change in seconds or minutes. And I don't want to fetching the prompt from your server for every call. What would be be better is for the prompts to be cached and provide a mechanism for looking for updates(for newer versions) and updating the cache. And not make this part of the execution pipeline.

### Additional information

Submitting as requested by Clemens from Langfuse based on our conversation on Intercom!

**Comment by marcklingen** [comment]
Thanks for the suggestion!

Current behavior: prompts are refetched (blocking) after the configured cache TTL, they are not fetched on every call

Suggested behavior: prompt are refetched after configured cache ttl but stale cache is served. I'd probably rather not have prompts refetch after first use on scheduled interval in background, this has the risk of refetching prompts which are not frequently used

What do you think about this @ajram23?

**Comment by ajram23** [comment]
@marcklingen To be honest, cache seems like a hack when you have versions. Curious why would you refetch prompts which are not frequently used if their versions don't change? May be I am missing something.

**Reply by marcklingen** [reply]
Currently the isActive flag is mostly used to fetch a prompt, we'll introduce multiple different labels (similar to tags of containers) this week to e.g. have a prod and a staging label. For these labels there's a need to refresh them to check if a the version changed and currently it seems faster to just fetch the latest version (incl the prompt) instead of just checking whether we still have the latest version cached. What do you think?

**Reply by ajram23** [reply]
Perhaps its best I describe my use case and provide me concerns instead of prescribing a solution. In my case I am only referring to production prompts and my concern with fetching every time is delaying/introducing another point of failure while answering a user. Label in this case will be production and it will always be active. The only instance the prompt will change is as I get feedback/score and find areas of brittle responses where I have to tweak the prompt (after eval) & deploy a new version. Hope that helps provide you with the context to come up with a right solve.

**Reply by marcklingen** [reply]
Thank you, short term you could add a very long TTL. Improvement: add "infinite" cache TTL?

**Reply by ajram23** [reply]
perhaps. I am hoping for a solution where I don't need to deploy for tweaking or iterating on prompts. and thank you for your engaging and congratulations on your week 1 of launch!

**Reply by marcklingen** [reply]
> perhaps. I am hoping for a solution where I don't need to deploy for tweaking or iterating on prompts. and thank you for your engaging and congratulations on your week 1 of launch!

Sorry, then I did not understand you correctly. With the SDK-level cache, Langfuse does not "fetching every time is delay/introducing another point of failure while answering a user". What do you suggest to change here?

> thank you for your engaging and congratulations on your week 1 of launch!

Thanks and no problem! I want to make prompt management in Langfuse awesome and any feedback is highly appreciated

**Reply by ajram23** [reply]
I made a suggestion at the beginning of the thread but I am empathetic to your other concerns. At the end of the day, want to be able to deploy the latest prompts to the user with minimal mucking around and performance impact. For now I am going with once a day as a cache. BTW one related question - generations always seem 0, do I have to trigger something for them to increment?

**Reply by marcklingen** [reply]
> For now I am going with once a day as a cache

Sounds good


> BTW one related question - generations always seem 0, do I have to trigger something for them to increment?

you'd need to link traces to prompt versions, see prompt management docs for details on this as it depends on how you integrate for tracing in langfuse

**Reply by marcklingen** [reply]
@ajram23 FYI, we have just released this change. You can now reduce the cache TTL to, for example, minutes without sacrificing performance.

https://langfuse.com/changelog/2024-09-04-prompt-management-zero-latency

let me know if you have any thoughts/feedback! thanks again for raising this here

**Reply by ajram23** [reply]
@marcklingen Fantastic! Will try it out later this week and get back to you!

**Reply by ajram23** [reply]
@marcklingen Did you guys update packaging? Any chance you can go to <24.0? 
```
langchain-core 0.1.52 requires packaging>=23.2 and <24.0
langfuse 2.46.0 requires packaging>=24.1 and <25.0
```

**Reply by maxdeichmann** [reply]
https://github.com/langfuse/langfuse-python/releases/tag/v2.46.1
@ajram23  thanks for raising this. I just downgraded packaging.

**Reply by maxdeichmann** [reply]
@ajram23 i updated our dependencies so that you can use our SKD with `packaging = ">=23.2,<25.0"`

**Reply by ajram23** [reply]
@maxdeichmann Thanks for the quick fix! My pipeline seems to be working so I am assuming you fix is as well.

**Reply by marcklingen** [reply]
awesome! let me know if the new prompt caching works as expected in your environment, until now, I got a lot of positive feedback on this

**Reply by ajram23** [reply]
@marcklingen So far so good. Glad that you are! I must be submitting good feature requests then ;)

**Reply by marcklingen** [reply]
>  I must be submitting good feature requests then ;)

yes, thank you!

offtopic: do you have a new public demo of your application by the way? loved the demo you once showed me

**Reply by ajram23** [reply]
@marcklingen not yet, packaging on a Mac sucks! Fighting that battle now, will reach out once its done!

**Reply by marcklingen** [reply]
best of luck, i can imagine
